### PR TITLE
`Programming exercises`: Enable online code editor in read-only mode

### PR DESCRIPTION
**Pull request by coolchock**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This pull request implements a feature to enable the online code editor in read-only mode if the offline IDE is enabled
resolves https://github.com/ls1intum/Artemis/issues/7886

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a Programming Exercise with the online code editor enabled in the read-only mode. It should be possible to turn on read-only mode only if the offline IDE is enabled.
4. Navigate to the created exercise
5. Start the exercise
6. Open the code editor and check if it's read-only

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Programming Exercise

1. Find an exam with a Programming Exercise and set the option for the online editor in the read-only mode. It should be possible to turn on read-only mode only if the offline IDE is enabled.
2. Participate in the exam as a student and go to the Programming Exercise
3. Open the code editor and check if it's read-only

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced a read-only mode for the online code editor in programming exercises, enhancing flexibility in how exercises can be configured and participated in.
- **Enhancements**
	- Improved validation for programming exercise settings to ensure consistency between participation modes.
	- Updated user interface components to reflect the new read-only mode option, including visibility adjustments and warning messages in the online code editor.
- **Bug Fixes**
	- Fixed an issue where the read-only mode could be enabled without meeting necessary conditions, ensuring a smoother user experience.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 1]
```diff
@@ -63,6 +63,9 @@ public String getType() {
     @Column(name = "allow_online_editor", table = "programming_exercise_details")
     private Boolean allowOnlineEditor;
 
+    @Column(name = "allow_online_editor_read_only_mode", table = "programming_exercise_details")
+    private Boolean allowOnlineEditorReadOnlyMode;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:67]:
```suggestion
    private boolean allowOnlineEditorReadOnlyMode = false;
```
New `boolean` attributes should be non-nullable if possible.

(related open issue #6079)

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 2]
```diff
@@ -269,6 +272,14 @@ public void setAllowOnlineEditor(Boolean allowOnlineEditor) {
         this.allowOnlineEditor = allowOnlineEditor;
     }
 
+    public Boolean isAllowOnlineEditorReadOnlyMode() {
+        return allowOnlineEditorReadOnlyMode;
+    }
+
+    public void setAllowOnlineEditorReadOnlyMode(Boolean allowOnlineEditorReadOnlyMode) {
+        this.allowOnlineEditorReadOnlyMode = allowOnlineEditorReadOnlyMode;
+    }
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:281]:
```suggestion
    public boolean isAllowOnlineEditorReadOnlyMode() {
        return allowOnlineEditorReadOnlyMode;
    }

    public void setAllowOnlineEditorReadOnlyMode(boolean allowOnlineEditorReadOnlyMode) {
        this.allowOnlineEditorReadOnlyMode = allowOnlineEditorReadOnlyMode;
    }
```

**Review by Gusti2010 — DISMISSED** [review]
Manually tested on ts11_Legacy. Works as expected!

**Review by georgyia — DISMISSED** [review]
manually tested on ts1, works as expected

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 3]
```diff
@@ -264,6 +264,11 @@ public ResponseEntity<ProgrammingExercise> updateProgrammingExercise(@RequestBod
             return ResponseEntity.badRequest().headers(HeaderUtil.createAlert(applicationName,
                     "You need to allow at least one participation mode, the online editor or the offline IDE", "noParticipationModeAllowed")).body(null);
         }
+        if (Boolean.TRUE.equals(updatedProgrammingExercise.isAllowOnlineEditorReadOnlyMode())
+                && (Boolean.FALSE.equals(updatedProgrammingExercise.isAllowOnlineEditor()) || Boolean.FALSE.equals(updatedProgrammingExercise.isAllowOfflineIde()))) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:268]:
```suggestion
        if (updatedProgrammingExercise.isAllowOnlineEditorReadOnlyMode()
                && (Boolean.FALSE.equals(updatedProgrammingExercise.isAllowOnlineEditor()) || Boolean.FALSE.equals(updatedProgrammingExercise.isAllowOfflineIde()))) {
```

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 4]
```diff
@@ -112,7 +112,12 @@
             "workdir": "Verzeichnis",
             "allowOnlineEditor": {
                 "title": "Online-Editor erlauben",
-                "alert": "Es muss mindestens eine Option (Offline-IDE oder Online-Editor) ausgewählt sein"
+                "alert": "Es muss mindestens eine Option (Offline-IDE oder Online-Editor) ausgewählt sein",
+                "readOnlyMode": {
+                    "title": "schreibgeschützter Modus",
+                    "alert": "Der schreibgeschützter Modus kann nur aktiviert werden, wenn die Offline-IDE aktiviert ist.",
```
Comment (b-fein) [inline src/main/webapp/i18n/de/programmingExercise.json:118]:
```suggestion
                    "alert": "Der schreibgeschützte Modus kann nur aktiviert werden, wenn die Offline-IDE aktiviert ist.",
```

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 5]
```diff
@@ -124,7 +124,12 @@
             "script": "Script",
             "allowOnlineEditor": {
                 "title": "Allow Online Editor",
-                "alert": "At least one option (Offline IDE or Online Editor) must be selected"
+                "alert": "At least one option (Offline IDE or Online Editor) must be selected",
+                "readOnlyMode": {
+                    "title": "read-only mode",
+                    "alert": "Read-only mode can only be enabled if the Offline IDE is enabled",
+                    "info": "For this exercise the online code editor is only enabled in the read-only mode"
```
Comment (b-fein) [inline src/main/webapp/i18n/en/programmingExercise.json:131]:
```suggestion
                    "alert": "Read-only mode can only be enabled if the Offline IDE is enabled.",
                    "info": "For this exercise the online code editor is only enabled in read-only mode."
```

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 6]
```diff
@@ -112,7 +112,12 @@
...
+                    "info": "Für diese Aufgabe ist der Online-Editor nur im schreibgeschützten Modus aktiviert"
```
Comment (b-fein) [inline src/main/webapp/i18n/de/programmingExercise.json:119]:
```suggestion
                    "info": "Für diese Aufgabe ist der Online-Editor nur im schreibgeschützten Modus aktiviert."
```

File: src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.ts [hunk 7]
```diff
@@ -17,4 +17,14 @@ export class ProgrammingExerciseDifficultyComponent {
     protected readonly ProjectType = ProjectType;
 
     faQuestionCircle = faQuestionCircle;
+
+    updateReadOnly() {
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.ts:21]:
```suggestion
    updateReadOnly() {
        this.programmingExercise.allowOnlineEditorReadOnlyMode = !this.isReadOnlyDisabled();
    }
```
The name `updateReadOnly` doesn't make sense to me if the flag can only be set to `false`. I'd expect the behavior of my suggestion.
If that's not correct consider renaming the method.

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.ts (hunk 7 above)
Comment (coolchock) [reply src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.ts]: Thanks for your feedback. The behavior that you suggested is not entirely what I wanted. The idea behind this function is to remove the check from the read-only option when conditions aren't satisfied anymore. The read-only checkbox can only be checked if both offline IDE and online editor are enabled. I renamed the function to make it more descriptive.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java (hunk 3 above)
Comment (coolchock) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java]: done

File: src/main/webapp/i18n/de/programmingExercise.json (hunk 4 above)
Comment (coolchock) [reply src/main/webapp/i18n/de/programmingExercise.json]: done

File: src/main/webapp/i18n/en/programmingExercise.json (hunk 5 above)
Comment (coolchock) [reply src/main/webapp/i18n/en/programmingExercise.json]: done

File: src/main/webapp/i18n/de/programmingExercise.json (hunk 6 above)
Comment (coolchock) [reply src/main/webapp/i18n/de/programmingExercise.json]: done

**Review by Predixx — DISMISSED** [review]
tested on ts1. lgtm!

**Review by nata-81 — DISMISSED** [review]
Tested on ts1, works as expected. However, I find it unnecessary to have the submit, refresh, create file, or directory buttons in read-only mode. Additionally, I believe it is unfitting for students to work in the online editor mode during the exam, rather than the normal mode where they can view the problem statement in full size.

**Review by AntonGeTUM — COMMENTED** [review]
Tested on ts11, works fine for normal programming exercises but accessing the Exam management page causes an Internal server error and none of the exams are shown. It is still possible to participate in an exam, though.
image

**Review by b-fein — CHANGES_REQUESTED** [review]
Merge-conflict in database changelog and also quite a bit behind `develop`. => Currently not testable on the test servers.

**Review by MaximilianAnzinger — APPROVED** [review]
code

**Review by georgyia — APPROVED** [review]
reapprove

**Review by Gusti2010 — APPROVED** [review]
reapprove

File: src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.html [hunk 8]
```diff
@@ -50,6 +59,21 @@
...
             </div>
+            <div class="form-check" style="margin-left: 20px">
+                <label [class]="'form-check-label'" [ngStyle]="shouldDisableReadOnlyMode() ? { color: 'grey' } : {}" for="field_allowOnlineEditorReadOnlyMode">
+                    <input
+                        class="form-check-input"
+                        type="checkbox"
+                        name="allowOnlineEditorReadOnlyMode"
+                        id="field_allowOnlineEditorReadOnlyMode"
+                        [(ngModel)]="programmingExercise.allowOnlineEditorReadOnlyMode"
+                        [disabled]="shouldDisableReadOnlyMode()"
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-difficulty.component.html:70]: No function-calls in templates. Consider reworking the logic or using Signals.
See the general client guidelines for more information

**Review by julian-christl — COMMENTED** [review]
Looks good, I just have one violation of our coding guidelines.

@Maintainers if you want this to be merged asap, consider this an approval. The violation should then be fixed afterwards. The violation won't cause major performance issues as far as I can assess this.

**Review by b-fein — APPROVED** [review]
Thanks for taking the time to implement the feature request!

Code changes look good to me.

**Review by b-fein — APPROVED** [review]
Code re-approve.

**Comment by b-fein** [comment]
Again, thanks for working on this issue.

Discussion on Slack with @coolchock and @krusche: we do not follow this approach, but re-use the read-only view that has been implemented for LocalVCS in the meantime also for GitLab setups.

(Discussion summary also added as comment to the issue in https://github.com/ls1intum/Artemis/issues/7886#issuecomment-2045965538.)

**Pull request closed without merging**